File: packages/ts/src/cnpj.ts
import type { ValidationResult } from "./types";

/** Machine-readable reasons a CNPJ can be invalid. */
export type CnpjReason =
  | "invalid-characters"
  | "invalid-length"
  | "repeated-digits"
  | "invalid-check-digit";

const CNPJ_LENGTH = 14;
const MASK = /[.\-/\s]/g;
const BASE_CHARS = /^[0-9A-Z]{12}$/;
const DV_CHARS = /^[0-9]{2}$/;
const ALL_SAME = /^(.)\1{13}$/;

// Official RFB weights (modulo 11). The first check digit is computed over the
// 12 base characters; the second over the 12 base characters plus the first DV.
const DV1_WEIGHTS = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2] as const;
const DV2_WEIGHTS = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2] as const;

const ALPHANUM = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const NUMERIC = "0123456789";

/** Remove mask characters and upper-case letters, keeping base + check digits. */
export function strip(value: string): string {
  return value.replace(MASK, "").toUpperCase();
}

/**
 * RFB alphanumeric rule: the numeric value of a character is its ASCII code
 * minus 48 (`'0'` → 0 … `'9'` → 9, `'A'` → 17 … `'Z'` → 42).
 */
function charValue(char: string): number {
  return char.charCodeAt(0) - 48;
}

function checkDigit(chars: string, weights: readonly number[]): number {
  let sum = 0;
  for (let i = 0; i < weights.length; i++) {
    sum += charValue(chars[i] as string) * (weights[i] as number);
  }
  const mod = sum % 11;
  return mod < 2 ? 0 : 11 - mod;
}

/**
 * Validate a CNPJ and explain *why* it is invalid. Transparently accepts both
 * the legacy numeric format and the 2026 alphanumeric format (A–Z allowed in
 * the 12 base characters; the two check digits are always numeric). Accepts
 * masked or unmasked, upper- or lower-case input.
 */
export function validateDetailed(value: string): ValidationResult<CnpjReason> {
  const cleaned = strip(value);
  if (cleaned.length !== CNPJ_LENGTH) {
    return { valid: false, reason: "invalid-length" };
  }
  const base = cleaned.slice(0, 12);
  const dv = cleaned.slice(12);
  if (!BASE_CHARS.test(base) || !DV_CHARS.test(dv)) {
    return { valid: false, reason: "invalid-characters" };
  }
  if (ALL_SAME.test(cleaned)) {
    return { valid: false, reason: "repeated-digits" };
  }
  const d1 = checkDigit(base, DV1_WEIGHTS);
  const d2 = checkDigit(`${base}${d1}`, DV2_WEIGHTS);
  if (d1 !== Number(dv[0]) || d2 !== Number(dv[1])) {
    return { valid: false, reason: "invalid-check-digit" };
  }
  return { valid: true, reason: null };
}

/** `true` when `value` is a valid CNPJ (numeric or alphanumeric). */
export function isValid(value: string): boolean {
  return validateDetailed(value).valid;
}

/**
 * `true` when the CNPJ uses the alphanumeric format — i.e. at least one letter
 * appears in the 12 base characters. Structural check only; it does not
 * validate the check digits.
 */
export function isAlphanumeric(value: string): boolean {
  const cleaned = strip(value);
  if (cleaned.length !== CNPJ_LENGTH) {
    return false;
  }
  return /[A-Z]/.test(cleaned.slice(0, 12));
}

/** Format 14 characters as `00.000.000/0000-00`. Throws on malformed input. */
export function format(value: string): string {
  const c = strip(value);
  if (c.length !== CNPJ_LENGTH || !BASE_CHARS.test(c.slice(0, 12)) || !DV_CHARS.test(c.slice(12))) {
    throw new RangeError(`Invalid CNPJ: cannot format ${JSON.stringify(value)}`);
  }
  return `${c.slice(0, 2)}.${c.slice(2, 5)}.${c.slice(5, 8)}/${c.slice(8, 12)}-${c.slice(12)}`;
}

export interface GenerateOptions {
  /** Return the masked form `00.000.000/0000-00` instead of bare characters. */
  formatted?: boolean;
  /** Generate an alphanumeric CNPJ (A–Z in the 12 base characters). */
  alphanumeric?: boolean;
}

/**
 * Generate a syntactically valid CNPJ. **For tests and fixtures only.** Pass
 * `{ alphanumeric: true }` to exercise the 2026 format.
 */
export function generate(options: GenerateOptions = {}): string {
  const pool = options.alphanumeric ? ALPHANUM : NUMERIC;
  let base = "";
  for (let i = 0; i < 12; i++) {
    base += pool[Math.floor(Math.random() * pool.length)];
  }
  if (new Set(base).size === 1) {
    base = (base[0] === "0" ? "1" : "0") + base.slice(1);
  }
  const d1 = checkDigit(base, DV1_WEIGHTS);
  const full = `${base}${d1}${checkDigit(`${base}${d1}`, DV2_WEIGHTS)}`;
  return options.formatted ? format(full) : full;
}

File: packages/py/src/br_kit/cnpj.py
"""CNPJ validation with full alphanumeric support. Mirrors ``cnpj.ts``.

Supports both the legacy numeric format and the 2026 alphanumeric format
(IN RFB 2.229/2024): A-Z allowed in the 12 base characters, the two check
digits always numeric, computed over each character's ASCII value minus 48.
"""

from __future__ import annotations

import random
import re

from ._types import ValidationResult

CNPJ_LENGTH = 14
_MASK = re.compile(r"[.\-/\s]")
_BASE_CHARS = re.compile(r"^[0-9A-Z]{12}$")
_DV_CHARS = re.compile(r"^[0-9]{2}$")
_ALL_SAME = re.compile(r"^(.)\1{13}$")

# Official RFB weights (modulo 11).
_DV1_WEIGHTS = (5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2)
_DV2_WEIGHTS = (6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2)

_ALPHANUM = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
_NUMERIC = "0123456789"


def strip(value: str) -> str:
    """Remove mask characters and upper-case letters, keeping base + check digits."""
    return _MASK.sub("", value).upper()


def _char_value(char: str) -> int:
    # RFB alphanumeric rule: numeric value is the ASCII code minus 48
    # ('0' -> 0 ... '9' -> 9, 'A' -> 17 ... 'Z' -> 42).
    return ord(char) - 48


def _check_digit(chars: str, weights: tuple[int, ...]) -> int:
    total = sum(_char_value(chars[i]) * weights[i] for i in range(len(weights)))
    mod = total % 11
    return 0 if mod < 2 else 11 - mod


def validate_detailed(value: str) -> ValidationResult:
    """Validate a CNPJ (numeric or alphanumeric) and explain *why* it is invalid.

    Accepts masked or unmasked, upper- or lower-case input.
    """
    cleaned = strip(value)
    if len(cleaned) != CNPJ_LENGTH:
        return ValidationResult(False, "invalid-length")
    base, dv = cleaned[:12], cleaned[12:]
    if not _BASE_CHARS.match(base) or not _DV_CHARS.match(dv):
        return ValidationResult(False, "invalid-characters")
    if _ALL_SAME.match(cleaned):
        return ValidationResult(False, "repeated-digits")
    d1 = _check_digit(base, _DV1_WEIGHTS)
    d2 = _check_digit(f"{base}{d1}", _DV2_WEIGHTS)
    if d1 != int(dv[0]) or d2 != int(dv[1]):
        return ValidationResult(False, "invalid-check-digit")
    return ValidationResult(True, None)


def is_valid(value: str) -> bool:
    """Return ``True`` when ``value`` is a valid CNPJ (numeric or alphanumeric)."""
    return validate_detailed(value).valid


def is_alphanumeric(value: str) -> bool:
    """Return ``True`` when at least one letter appears in the 12 base characters.

    Structural check only; it does not validate the check digits.
    """
    cleaned = strip(value)
    if len(cleaned) != CNPJ_LENGTH:
        return False
    return any("A" <= c <= "Z" for c in cleaned[:12])


def format(value: str) -> str:  # noqa: A001 - mirrors the cross-language API name
    """Format 14 characters as ``00.000.000/0000-00``. Raises on malformed input."""
    c = strip(value)
    if len(c) != CNPJ_LENGTH or not _BASE_CHARS.match(c[:12]) or not _DV_CHARS.match(c[12:]):
        raise ValueError(f"Invalid CNPJ: cannot format {value!r}")
    return f"{c[:2]}.{c[2:5]}.{c[5:8]}/{c[8:12]}-{c[12:]}"


def generate(*, formatted: bool = False, alphanumeric: bool = False) -> str:
    """Generate a syntactically valid CNPJ. **For tests and fixtures only.**

    Pass ``alphanumeric=True`` to exercise the 2026 format.
    """
    pool = _ALPHANUM if alphanumeric else _NUMERIC
    base = "".join(random.choice(pool) for _ in range(12))
    if len(set(base)) == 1:
        base = ("1" if base[0] == "0" else "0") + base[1:]
    d1 = _check_digit(base, _DV1_WEIGHTS)
    full = f"{base}{d1}{_check_digit(f'{base}{d1}', _DV2_WEIGHTS)}"
    return format(full) if formatted else full
